Data/Primitive/UnliftedArray.hs

```diff
@@ -137,6 +141,22 @@ instance PrimUnlifted (MV.MutVar s a) where
   toArrayArray# (MV.MutVar mv#) = unsafeCoerce# mv#
   fromArrayArray# aa# = MV.MutVar (unsafeCoerce# aa#)
 
+instance PrimUnlifted (MVar a) where
+  toArrayArray# (MVar mv#) = unsafeCoerce# mv#
```

Comment:
If you look at my `ArrayArray#` wrapping stuff, you'll see the same thing. The primop API is that `ArrayArray#` stores values of type `ArrayArray#`, and you `unsafeCoerce` between that and the type you actually want. I agree it's not a great API, but it's what we have right now.

In the future, maybe there will be an array type that is polymorphic over unlifted types that could be more safe at the primop level. But right now the API I wrote is the safe wrapper around the unsafe API.